### PR TITLE
Fixes multiple runtime issues

### PR DESCRIPTION
**Pull request by timonmentink**
Fixed multiple issues in this PR.

I have ros2-dotnet now building & working examples under Windows 10 + Humble & Ubuntu 22.04 + Humble.

There where some typo`s, missing functions in _native.dll and old dll loading functions.

### CODE REVIEW

**Comment by hoffmann-stefan** [comment]
@timonmentink Thanks for the contributions 👍 Currently somewhat busy, but hope to get to review this soon.

**Comment by timonmentink** [comment]
Seems like pre Humble uses python_executable. Created a list that contains all the newer distro`s for easy checks

File: rcldotnet_common/DllLoadUtils.cs [hunk 1]
```diff
@@ -231,10 +231,10 @@ public class DllLoadUtilsWindowsDesktop : DllLoadUtilsAbs
             [DllImport("kernel32.dll", EntryPoint = "LoadLibraryA", SetLastError = true, ExactSpelling = true)]
             private static extern IntPtr LoadLibraryA(string fileName, int reserved = 0);
 
-            [DllImport("kernel32.dll", SetLastError = true, ExactSpelling = true)]
+            [DllImport("kernel32.dll", EntryPoint = "FreeLibrary", SetLastError = true, ExactSpelling = true)]
```
Comment (hoffmann-stefan) [inline rcldotnet_common/DllLoadUtils.cs:234]: Seems this was broken by this commit: https://github.com/ros2-dotnet/ros2_dotnet/commit/3bd4a55dd795c413e563c88cae9381ef1dc41f5d

Could you add the `EntryPoint` for UWP here as well?

File: rosidl_generator_dotnet/cmake/rosidl_generator_dotnet_generate_interfaces.cmake [hunk 2]
```diff
@@ -35,6 +35,15 @@ set(_output_path
 set(_generated_cs_files "")
 set(_generated_c_ts_files "")
 set(_generated_h_files "")
+set(ros2_distro "$ENV{ROS_DISTRO}")
+set(humble_distro_and_newer humble iron rolling)
```
Comment (hoffmann-stefan) [inline rosidl_generator_dotnet/cmake/rosidl_generator_dotnet_generate_interfaces.cmake:39]: I think we should invert the logic here:
This list would need to be extend going forward, but if we have a list of older ROS2 distros (I think `foxy` and `galacic` should be enougth) this won't break if j-turtle gets released.

File: rosidl_generator_dotnet/cmake/rosidl_generator_dotnet_generate_interfaces.cmake [hunk 3]
```diff
@@ -215,9 +225,13 @@ if(_generated_c_ts_files)
...
-
-  rosidl_target_interfaces(${_target_name}
-    ${rosidl_generate_interfaces_TARGET} rosidl_typesupport_c)
+  
+  if(ros2_distro IN_LIST humble_distro_and_newer)
+    rosidl_get_typesupport_target(c_typesupport_target "${rosidl_generate_interfaces_TARGET}" "rosidl_typesupport_c")
+    target_link_libraries(${_target_name} "${c_typesupport_target}")
+  else()
+    rosidl_target_interfaces(${_target_name} ${rosidl_generate_interfaces_TARGET} rosidl_typesupport_c)
+  endif()
```
Comment (hoffmann-stefan) [inline rosidl_generator_dotnet/cmake/rosidl_generator_dotnet_generate_interfaces.cmake:234]: Not sure if I got this right, could you explain why this change is necessary?
Seems like the way interfaces are generated changed in humble?

File: rosidl_generator_dotnet/cmake/rosidl_generator_dotnet_generate_interfaces.cmake [hunk 4]
```diff
@@ -35,6 +35,15 @@ set(_output_path
...
 set(_generated_c_ts_files "")
 set(_generated_h_files "")
+set(ros2_distro "$ENV{ROS_DISTRO}")
+set(humble_distro_and_newer humble iron rolling)
+
+
+if(ros2_distro IN_LIST humble_distro_and_newer) 
+set(PYTHON_COMMAND ${PYTHON_EXECUTABLE})
+else
+set(PYTHON_COMMAND Python3::Interpreter)
```
Comment (hoffmann-stefan) [inline rosidl_generator_dotnet/cmake/rosidl_generator_dotnet_generate_interfaces.cmake:45]: I get a parse error for this line on both `foxy` and `humble` on my machine (two different vm's for both ubuntu versions):

```
~/humble/ros2_dotnet_ws$ cat log/latest_build/unique_identifier_msgs/stdout_stderr.log
-- The C compiler identification is GNU 11.3.0
-- The CXX compiler identification is GNU 11.3.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found ament_cmake: 1.3.4 (/opt/ros/humble/share/ament_cmake/cmake)
-- Found Python3: /usr/bin/python3.10 (found version "3.10.6") found components: Interpreter
-- Found rosidl_default_generators: 1.2.0 (/home/stefan/humble/ros2_dotnet_ws/install/rosidl_default_generators/share/rosidl_default_generators/cmake)
-- Using all available rosidl_typesupport_c: rosidl_typesupport_fastrtps_c;rosidl_typesupport_introspection_c
-- Found rosidl_adapter: 3.1.4 (/opt/ros/humble/share/rosidl_adapter/cmake)
-- Using all available rosidl_typesupport_cpp: rosidl_typesupport_fastrtps_cpp;rosidl_typesupport_introspection_cpp
-- Found ament_cmake_ros: 0.10.0 (/opt/ros/humble/share/ament_cmake_ros/cmake)
-- Using all available rosidl_typesupport_c: rosidl_typesupport_fastrtps_c;rosidl_typesupport_introspection_c
-- Using all available rosidl_typesupport_cpp: rosidl_typesupport_fastrtps_cpp;rosidl_typesupport_introspection_cpp
CMake Error at /home/stefan/humble/ros2_dotnet_ws/install/rosidl_generator_dotnet/share/rosidl_generator_dotnet/cmake/rosidl_generator_dotnet_generate_interfaces.cmake:45:
  Parse error.  Expected "(", got newline with text "

  ".
Call Stack (most recent call first):
  /opt/ros/humble/share/ament_cmake_core/cmake/core/ament_execute_extensions.cmake:48 (include)
  /opt/ros/humble/share/rosidl_cmake/cmake/rosidl_generate_interfaces.cmake:286 (ament_execute_extensions)
  CMakeLists.txt:21 (rosidl_generate_interfaces)
```

on my humble vm i use `cmake version 3.22.1`, on foxy `cmake version 3.16.3`

Why did you change this? Do you have any clue why this would break on my machine but not on CI?
Can look at this again later on, but maybe you already know something and can give some pointers :)

**Review by hoffmann-stefan — CHANGES_REQUESTED** [review]
Finally had time to look at this.
Did try to build this on my machine, but both foxy and humble did not build, see comment below.

Have some other small changes and questions as well.

Thanks for your contribution, seems like you have got ros2_dotnet to work on windows again :)

**Comment by timonmentink** [comment]
After a long time, i hopefully fixed the issues :)

File: rosidl_generator_dotnet/cmake/rosidl_generator_dotnet_generate_interfaces.cmake (hunk 4 above)
Comment (timonmentink) [reply rosidl_generator_dotnet/cmake/rosidl_generator_dotnet_generate_interfaces.cmake]: i forgot the () at line 44, was else but needed to be else()

File: rosidl_generator_dotnet/cmake/rosidl_generator_dotnet_generate_interfaces.cmake (hunk 3 above)
Comment (timonmentink) [reply rosidl_generator_dotnet/cmake/rosidl_generator_dotnet_generate_interfaces.cmake]: Yeah is indeed the deprecation of that function

File: rosidl_generator_dotnet/cmake/rosidl_generator_dotnet_generate_interfaces.cmake (hunk 2 above)
Comment (timonmentink) [reply rosidl_generator_dotnet/cmake/rosidl_generator_dotnet_generate_interfaces.cmake]: I inverted the logic on the last commit

File: rcldotnet_common/DllLoadUtils.cs (hunk 1 above)
Comment (timonmentink) [reply rcldotnet_common/DllLoadUtils.cs:234]: fixed this :)

**Review by hoffmann-stefan — APPROVED** [review]
works locally on my machine, CI is happy to. Looks good 👍
Thanks for your contribution :)

**Comment by hoffmann-stefan** [comment]
> After a long time, i hopefully fixed the issues :)

I have been busy as well for some time. Became a father some weeks ago, so my spare time for contributing here was blocked by preparing things at home ;)

> I have ros2-dotnet now building & working examples under Windows 10 + Humble & Ubuntu 22.04 + Humble.

Seems like you are the first one to get this working with humble on windows (who published this online). Could you comment on [this issue](https://github.com/ros2-dotnet/ros2_dotnet/issues/80#issuecomment-1518545762) about it? Are there still some workarounds needed or missing documentation to get it working on windows? If it should work out of the box from your perspective I would close that issue.
